homeassistant/components/private_ble_device/coordinator.py

```diff
@@ -0,0 +1,236 @@
+"""Central manager for tracking devices with random but resolvable MAC addresses."""
+from __future__ import annotations
+
+from collections.abc import Callable
+import logging
+from typing import cast
+
+from bluetooth_data_tools import get_cipher_for_irk, resolve_private_address
+from cryptography.hazmat.primitives.ciphers import Cipher
+
+from homeassistant.components import bluetooth
+from homeassistant.components.bluetooth.match import BluetoothCallbackMatcher
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+UnavailableCallback = Callable[[bluetooth.BluetoothServiceInfoBleak], None]
+Cancellable = Callable[[], None]
+
+
+def async_last_service_info(
+    hass: HomeAssistant, irk: bytes
+) -> bluetooth.BluetoothServiceInfoBleak | None:
+    """Find a BluetoothServiceInfoBleak for the irk.
+
+    This iterates over all currently visible mac addresses and checks them against `irk`.
+    It returns the newest.
+    """
+
+    # This can't use existing data collected by the coordinator - its called when
+    # the coordinator doesn't know about the IRK, so doesn't optimise this lookup.
+
+    cur: bluetooth.BluetoothServiceInfoBleak | None = None
+    cipher = get_cipher_for_irk(irk)
+
+    for service_info in bluetooth.async_discovered_service_info(hass, False):
```

Comment:
You could likely exclude all the ones that aren't private/random address types

Reply:
I guess we don't expose the bit that tells if its random or public in a meaningful way.

Reply:
https://github.com/home-assistant/core/blob/f48e8623da8443f72e617db5d22f919a28a8c063/homeassistant/components/esphome/bluetooth/scanner.py#L46

Reply:
We kick out any that aren't RPA in the resolve address function. But I would like to plumb the address type into the Bluetooth component to avoid the callbacks in the first place.

Reply:
That could be quite invasive tho.

Reply:
maybe we should standardize that

Reply:
Yeah. I was thinking all the matchers in the Bluetooth integration could be expanded to have an address_type filter which would be a list of a new AddressType enum.

Doing it there would mean that we could ignore AddressType.RANDOM_UNRESOLVABLE for most integrations (except iBeacon).

We could also kick out unresolvable and resolvable from the ble tracker integration, as it can't use them.

I could look at that in a follow up?

Reply:
Like you say, would need to expose the static/random bit everywhere to do it properly to do this.

Reply:
I think ^ is outside scope of this PR, but definitely want to come back to it as like I say I think it could be a useful filter in other integration as well.

Reply:
Its definitely out of scope. We would have an UNKNOWN value for the first iteration as well since not all scanners send that data all the way through right now